keep.tex

```diff
@@ -530,38 +531,188 @@ \subsection{Concerns with active attacks}
-in the keep market. Requiring a native token, rather than the
-underlying blockchain's currency, means providers will suffer from
-negative externalities in the presence of malicious behavior.
+There are a few design considerations for such a system
+\begin{itemize}
+  \item Providers can't be allowed an unfair advantage over each other
+      in the node selection process.
+  \item Each block on the public chain will require at least one
+      random number of sufficient size. Today's Ethereum block time is
+      25 seconds, but that will likely change significantly in the
+      future. The RNG process needs to be fast enough to support much
+      shorter block time, if necessary.
+  \item RNG needs to be resilient to node failure. Failure in
+      production means no new keeps can be created, so resilience
+      to partitions between providers as well as against active denial
+      of service attacks is desirable.
+  \item While not a hard system requirement, providing the Ethereum
+      network with a trusted source of randomness will also be a great
+      boon to other projects.
+\end{itemize}
+
+Most distributed key generation schemes are too slow or prone to
+manipulation to be considered. Any scheme we choose should provide
+good performance, regardless of the number of particating providers.
+Instead, most generation schemes require rounds of communication
+between participants, slowing down the key generation process and
+providing a large surface for communication failure.
 
-Keep operations can be paid for with the native token, or with the
-underlying blockchain's currency, based on the keep market.
+Fortunately, the Dfinity team has solved these issues with their
+random beacon design, based on a concept they call threshold relay
+\cite{thresholdRelay}.
+
+\subsubsection{Threshold relay}
+
+\begin{table*}[t]
+  \centering
+  \begin{tabular}{|rp{10cm}|}
+  \hline
+  \multicolumn{2}{|c|}{\textit{Iterative threshold signatures for
+  randomness on existing chains}} \\
+  \textbf{Registration:} & As providers join the network, they
+  register with at least one threshold group $G_i$ of all groups $G$,
+  generating a share of the group's private key, $s_i$. Threshold groups are
+  capped at $c$ members, and may intersect. Groups that have reached
+  this maximum size publish their public key to the blockchain. We'll
+  designate such groups as $G_{registered}$. \\
+  \textbf{Trusted setup:} & A trusted party posts a random value
+  $r_0$ to the blockchain as the beacon's first output. \\
+  \textbf{Bootstrapping:} & $mod(r_{0}, |G_{registered}|)$ is
+  used to select a registered threshold group, $G_i$, from
+  $G_{registered}$. $G_i$ signs $r_0$ and publishes the
+  result, $r_1 = threshold(r_0, s_{0\rightarrow{t}})$ where
+  $s_{0\rightarrow{t}}$ is the minimal shares necessary the the group
+  to produce a signature. \\
+  \textbf{Iteration:} & Each block published on the chain will include
```

Comment:
Normally, this would suffer from the last signer in a threshold group being able to bias the signature. Make it clear that we'll be using BLS or another deterministic signature